### PR TITLE
feat: implement OpenAPI details sections

### PR DESCRIPTION
**Pull request by zchsh**
## 🔗 Relevant links

- [Preview link](https://dev-portal-git-BRANCH_NAME-hashicorp.vercel.app/PATH_TO_VIEW) 🔎
- 🎟️ [Implement revised "PropertyDetails" component for "details" area](https://app.asana.com/0/1204678746647847/1205011720080299/f)
- 🎟️ [Implement operation "details" column for new API docs view](https://app.asana.com/0/1204678746647847/1205011720080302/f)

## 🗒️ What

- Implements the presentational `PropertyDetails` component
- Implements the server-side logic to build props for the details section, so that `PropertyDetails` can be shown in situ with real data

> **Note**: initially I'd thought this work might make sense as two separate PRs, but am now leaning towards keeping it in this one PR, as either of the above items felt difficult to really validate in the preview without the other. Happy to take a different approach though if it feels like it'd be worth it 🙇‍♂️

## 📸 Design Screenshots

🎨 [Design reference in Figma][figma]

| Before | After |
| - | - |
| listapps-before | listapps-after |
| createapp-before | createapp-after |

## 🧪 Testing

- [ ] Visit [/hcp/api-docs/vault-secrets]
    - All operations should show `Request` and `Response` sections, matching the design spec
    - The property details shown in each of the above sections should match the design spec
    - Information shown in the details section should have parity with information shown on [the same page upstream][upstream/hcp/api-docs/vault-secrets]

## 💭 Anything else?

Not at the moment! I think we probably could tighten up the spacing here, but I feel like that's probably better to leave to future PRs, and continue getting the "non-tweaks" stuff implemented first 👍

[figma]: https://www.figma.com/file/wjoC3ZVsA1kDuPsFLjEFuY/API-Template?type=design&node-id=526-34648&mode=design&t=jgysJrOUIjRdzakJ-4
[preview]: https://dev-portal-git-zsopen-api-property-details-hashicorp.vercel.app/
[/hcp/api-docs/vault-secrets]: https://dev-portal-git-zsopen-api-property-details-hashicorp.vercel.app/hcp/api-docs/vault-secrets
[upstream/hcp/api-docs/vault-secrets]: https://developer.hashicorp.com/hcp/api-docs/vault-secrets

### CODE REVIEW

File: src/views/open-api-docs-view/components/property-details/index.tsx [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+	nestedProperties?: PropertyDetailsProps[]
+	depth?: number
+}
+
+/**
+ * Render a top-level property, with optional nested properties.
+ *
+ * Future: will need to support "beta" badge for HCP Packer docs.
+ */
+export function PropertyDetails({
```
Comment (zchsh) [inline src/views/open-api-docs-view/components/property-details/index.tsx:30]: This component is the bulk of the presentational work here. `.tsx` and `.css` files here are a lil longer than usual, but leaned toward keeping them co-located in one file since they're pretty interdependent, and have the whole potentially recursive rendering thing going on.

File: src/views/open-api-docs-view/server.ts [hunk 2]
```diff
@@ -101,7 +101,7 @@ export async function getStaticProps({
 			: await fetchGithubFile(sourceFile)
 	const schemaData = await parseAndValidateOpenApiSchema(schemaFileString)
 	const { title } = schemaData.info
-	const operationProps = getOperationProps(schemaData)
+	const operationProps = await getOperationProps(schemaData)
```
Comment (zchsh) [inline src/views/open-api-docs-view/server.ts:104]: Had to update this file and a few other files to get the `async` flow going for markdown processing.

File: src/views/open-api-docs-view/types.ts [hunk 3]
```diff
@@ -6,19 +6,7 @@
...
- */
-export interface PropertyDetailProps {
-	name: string
-	/**
-	 * Some temporary data to mess around with during prototyping.
-	 * TODO: remove this for the production implementation.
-	 */
-	_placeholder: $TSFixMe
-}
+import type { PropertyDetailsProps } from './components/property-details'
```
Comment (zchsh) [inline src/views/open-api-docs-view/types.ts:9]: Have updated the types here to reflect the not-placeholder implementation.

File: src/views/open-api-docs-view/utils/get-operation-props.ts [hunk 4]
```diff
@@ -7,21 +7,17 @@
 import { snakeCase } from 'change-case'
 import slugify from 'slugify'
 // Local
-import {
-	getBodyParameterProps,
-	getPropertyDetailPropsFromParameter,
-	truncateHcpOperationPath,
-} from './'
+import { getRequestData, getResponseData, truncateHcpOperationPath } from './'
```
Comment (zchsh) [inline src/views/open-api-docs-view/utils/get-operation-props.ts:10]: I've split out `getRequestData` and `getResponseData` to separate functions to try to keep this file easy to skim.

File: src/views/open-api-docs-view/utils/get-property-detail-props.ts [hunk 5]
```diff
@@ -3,31 +3,120 @@
...
+	}
+}
+
+/**
+ * Given a schema object,
+ * Return a typeString summarizing the schema,
+ * descriptionHtml generated from the schema.description,
+ * and an array of nested properties with fully built props.
+ */
+async function getPropertyDetailsFromSchema(
```
Comment (zchsh) [inline src/views/open-api-docs-view/utils/get-property-detail-props.ts:56]: This is where the more interesting parsing happens, transforming parameters and schemas into the props we show on the front-end.

File: src/views/open-api-docs-view/utils/get-request-data.ts [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+} from './get-property-detail-props'
+// Types
+import type { PropertyDetailsProps } from '../components/property-details'
+import type { OpenAPIV3 } from 'openapi-types'
+
+/**
+ * Given OpenAPI parameter objects,
+ * Return request data formatted for display.
+ */
+export async function getRequestData(
```
Comment (zchsh) [inline src/views/open-api-docs-view/utils/get-request-data.ts:19]: A lot of the work in `getRequestData` is being thorough in respecting OpenAPIV3 types, and collecting the data we need from various parts of the schema.

File: src/views/open-api-docs-view/utils/get-response-data.ts [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+const RESPONSE_CODE_HEADINGS: Record<string, string> = {
+	'200': 'Successful Response',
+	default: 'Default Response',
+}
+
+/**
+ * Given OpenAPI responses data,
+ * Return response data formatted for display.
+ */
+export async function getResponseData(
```
Comment (zchsh) [inline src/views/open-api-docs-view/utils/get-response-data.ts:22]: As with `get-request-data`, a lot of the work in `getResponseData` is being thorough in respecting types, and parsing out the parts of the schema we want.

File: src/views/open-api-docs-view/components/property-details/index.tsx [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+					<span className={s.baseType}>{type}</span>
+					{isRequired ? <Badge text="Required" color="highlight" /> : null}
+				</div>
+				{description ? (
+					<div
+						className={s.baseDescription}
+						dangerouslySetInnerHTML={{ __html: description }}
+					/>
+				) : null}
+			</div>
```
Comment (kendallstrautman) [inline src/views/open-api-docs-view/components/property-details/index.tsx:52]: Screenshot 2023-08-02 at 11 33 58 AM
Screenshot 2023-08-02 at 11 35 01 AM

Makes sense if we don't want to do this style now, but just noting the discrepancy.

File: src/views/open-api-docs-view/components/operation-details/index.tsx [hunk 9]
```diff
@@ -3,19 +3,70 @@
...
+ * Render request and response data for an operation.
  */
-export function OperationDetails({ operation }: { operation: OperationProps }) {
+export function OperationDetails({
+	requestData,
+	responseData,
+}: {
+	requestData: { heading: string; propertyDetails: PropertyDetailsProps[] }[]
+	responseData: { heading: string; propertyDetails: PropertyDetailsProps[] }[]
+}) {
```
Comment (kendallstrautman) [inline src/views/open-api-docs-view/components/operation-details/index.tsx:18]: ⛏️ Since its such a long line length, thoughts on defining this in an interface above?

File: src/views/open-api-docs-view/components/property-details/index.tsx [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+import { MdxInlineCode } from 'components/dev-dot-content/mdx-components'
+import Badge from 'components/badge'
+import { IconCornerDownRight16 } from '@hashicorp/flight-icons/svg-react/corner-down-right-16'
+import s from './property-details.module.css'
+
+/**
+ * Types are shared by the "base" and "deeper" variations of
+ * the property details component.
+ */
+export type PropertyDetailsProps = {
```
Comment (kendallstrautman) [inline src/views/open-api-docs-view/components/property-details/index.tsx:16]: Any particular reason for using `type` instead of `interface` here?

File: src/views/open-api-docs-view/components/property-details/property-details.module.css [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+	list-style: none;
+	margin: 0;
+	padding: 0;
+	width: 100%;
+}
+
+/* Description styles markdown content. Note we do not use MDX in this
+   case, so specific styles we want for markdown content will need
+	 to be added here. */
+.description {
```
Comment (kendallstrautman) [inline src/views/open-api-docs-view/components/property-details/property-details.module.css:25]: Could we share some of the mdx base content styles?

File: src/views/open-api-docs-view/types.ts [hunk 12]
```diff
@@ -32,9 +20,14 @@ export interface OperationProps {
...
-	pathParameters?: PropertyDetailProps[]
-	bodyParameters?: PropertyDetailProps[]
+	requestData: {
+		heading: string
+		propertyDetails: PropertyDetailsProps[]
+	}[]
+	responseData: {
+		heading: string
+		propertyDetails: PropertyDetailsProps[]
+	}[]
```
Comment (kendallstrautman) [inline src/views/open-api-docs-view/types.ts:30]: Do you want to unify any types here between the interface used with the `OperationDetails` component?

File: src/views/open-api-docs-view/utils/get-operation-props.ts [hunk 13]
```diff
@@ -74,9 +57,8 @@ export function getOperationProps(
...
 				},
 				summary: operation.summary,
-				bodyParameters,
-				pathParameters,
-				queryParameters,
+				requestData,
+				responseData,
 				_placeholder: {
 					__type: type,
 					__path: path,
```
Comment (kendallstrautman) [inline src/views/open-api-docs-view/utils/get-operation-props.ts:64]: Still need this?

File: src/views/open-api-docs-view/utils/get-response-data.ts [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+
+/**
+ * Map known response codes to human-readable headings.
+ *
+ * If an response code isn't in this map, we use the response code itself
+ * as the heading.
+ */
+const RESPONSE_CODE_HEADINGS: Record<string, string> = {
+	'200': 'Successful Response',
+	default: 'Default Response',
```
Comment (kendallstrautman) [inline src/views/open-api-docs-view/utils/get-response-data.ts:15]: Not sure if something like this would be helpful https://github.com/prettymuchbryce/http-status-codes

**Review by kendallstrautman — APPROVED** [review]
Preview looks great! Just a few minor comments 💯

File: src/views/open-api-docs-view/components/property-details/index.tsx (hunk 8 above)
Comment (zchsh) [reply src/views/open-api-docs-view/components/property-details/index.tsx:52]: Ah yeah, I should have made an explicit note of this! For now we're going with static display rather than interactive form 👍

File: src/views/open-api-docs-view/components/property-details/index.tsx (hunk 10 above)
Comment (zchsh) [reply src/views/open-api-docs-view/components/property-details/index.tsx:16]: No particular reason, just slipped my mind 😅 Updated to `interface`! 👍

File: src/views/open-api-docs-view/components/property-details/index.tsx [hunk 15]
```diff
@@ -0,0 +1,129 @@
+/**
+ * Copyright (c) HashiCorp, Inc.
+ * SPDX-License-Identifier: MPL-2.0
+ */
+
+import classNames from 'classnames'
+import { MdxInlineCode } from 'components/dev-dot-content/mdx-components'
+import Badge from 'components/badge'
+import { IconCornerDownRight16 } from '@hashicorp/flight-icons/svg-react/corner-down-right-16'
```
[the remark this answers is not in the document]
Comment (zchsh) [reply src/views/open-api-docs-view/components/property-details/index.tsx:9]: Thanks for catching this, organized these imports a bit 👍 And yeah that'd be awesome - especially with Intellisense or whatever auto-adding imports, it'd be great to have em magically snap into a reasonable order.

File: src/views/open-api-docs-view/components/property-details/property-details.module.css (hunk 11 above)
Comment (zchsh) [reply src/views/open-api-docs-view/components/property-details/property-details.module.css:25]: I think it might be technically possible, but I'm a bit hesitant to, cause I feel like it might be a bit of an early abstraction... One notable difference with these styles is the base font-size varies depending on the nesting level of the property in question.

My intent for now was to keep the styles here isolated, especially since these descriptions seem to only contain basic text styling and links (as opposed to more fully-featured markdown use such as block elements). I think if we find ourselves reaching farther to style many other elements (say lists, code blocks, etc) then switching this to use `next-mdx-remote` would probably make sense 👍

File: src/views/open-api-docs-view/types.ts (hunk 12 above)
Comment (zchsh) [reply src/views/open-api-docs-view/types.ts]: Ah yeah, great call! 🙌 Have updated to do that 🙇‍♂️

File: src/views/open-api-docs-view/utils/get-response-data.ts (hunk 14 above)
Comment (zchsh) [reply src/views/open-api-docs-view/utils/get-response-data.ts]: Nice, yeah that's super helpful! Have updated to show `Default Response` if the OpenAPI response code is `default` (which isn't a supported code in `http-status-codes`), and otherwise grab `getReasonPhrase(responseCode)` using `http-status-codes` and combine that with the code itself for a more informative heading 👍

File: src/views/open-api-docs-view/components/property-details/index.tsx (hunk 8 above)
Comment (braydenlove) [reply src/views/open-api-docs-view/components/property-details/index.tsx:56]: Sorry @kendallstrautman I should have made that more clear. Future ideas ⭐

**Comment by braydenlove** [comment]
@zchsh a few comments that I will leave individually.

Curious if we are able to add anchor links to the top level parameters so folks can share specific sections of the site. If not they do have the ability to link sections based on the anchors in the sidebar.

**Comment by braydenlove** [comment]
Curious if we are able to shorten this to HCP instead of HashiCorp Cloud Platform, I realize this is author-able! Icon with the multi-line <li> text isn't ideal but we can and should support it. Mostly a nit as it looks bad. 
image

**Comment by braydenlove** [comment]
@zchsh I think we can also remove the `<hr>` as they aren't feeling useful to me personally. LMK if you disagree.
image

**Comment by braydenlove** [comment]
Bottom of the page needs some spacing as it ending at the footer. I would just use `128px` like we do on the other docs pages. Do we want to add an edit this page on GitHub link or some sort of feedback mechanism? 
image

**Comment by braydenlove** [comment]
When you scroll down to a section the top `<li>` HashiCorp Cloud Platform shouldn't have the active styling anymore. Ideally it could be used to be a back to top tool 🤞🏼 
image

**Comment by zchsh** [comment]
@braydenlove Thanks for all the awesome feedback! 🙌 💯 Captured most of it in separate small tasks I'll follow up with:
- 🎟️ [Add anchor links to top-level "PropertyDetails" parameters](https://app.asana.com/0/1204678746647847/1205211654360778/f)
- 🎟 [Update #operation-anchor in URL as page is scrolled](https://app.asana.com/0/1204678746647847/1205211659139651/f)
- 🎟 [Shorten sidebar text HashiCorp Cloud Platform to HCP somehow](https://app.asana.com/0/1204678746647847/1205211659139655/f)
- 🎟 [Add 128px padding to the bottom of the page](https://app.asana.com/0/1204678746647847/1205211659139659/f)
- 🎟 [Update top-level sidebar link to act as "scroll to top" link](https://app.asana.com/0/1204678746647847/1205211659139665/f)

Have update this PR to remove the `<hr />` as well, and ensure the gap there is set to `20px` 👍

Let me know if this looks good to merge! 🙇‍♂️

**Review by braydenlove — APPROVED** [review]
Looks good, thanks!
